File: src/main/java/gameoflife/algorithm/GameOfLife.java
package gameoflife.algorithm;

import java.util.Collection;
import java.util.Map;
import java.util.Optional;
import java.util.function.Predicate;
import java.util.stream.Collectors;
import java.util.stream.Stream;

public class GameOfLife {
    private final Map<String, Cell> liveCells;
    private final Predicate<Long> isReproducible = n -> n == 3;
    private final Predicate<Long> isNextGeneration = n -> 2 == n || n == 3;

    public GameOfLife(Map<String, Cell> liveCells) {
        this.liveCells = liveCells;
    }

    public GameOfLife tick() {
        return new GameOfLife(getNextLiveCellsMap());
    }

    public boolean isActive(int x, int y) {
        return isActive(CellKt.toString(x, y));
    }

    private boolean isActive(String key) {
        return liveCells.get(key) != null;
    }

    Collection<Cell> getLiveCells() {
        return Optional.of(liveCells)
                .filter(cells -> !cells.isEmpty())
                .map(Map::values)
                .orElseThrow(() -> new RuntimeException("No more living cells"));
    }

    private Map<String, Cell> getNextLiveCellsMap() {
        return Stream.concat(getNextGenerationCells(), getReproducibleCells())
                .collect(Collectors.toMap(Cell::toString, cell -> cell));
    }

    private Stream<Cell> getNextGenerationCells() {
        return filteredCells(getLiveCells().stream(), isNextGeneration);
    }

    private Stream<Cell> getReproducibleCells() {
        return filteredCells(getInactiveNeighbours(), isReproducible);
    }

    private Stream<Cell> filteredCells(Stream<Cell> cells, Predicate<Long> isActiveCell) {
        return cells.filter(cell -> isActiveCell.test(countActiveNeighbours(cell)));
    }

    private long countActiveNeighbours(Cell cell) {
        return cell.getNeighbours()
                .filter(c -> isActive(c.toString()))
                .count();
    }

    Stream<Cell> getInactiveNeighbours() {
        return getLiveCells().stream()
                .flatMap(Cell::getNeighbours)
                .distinct()
                .filter(this::isNotActive);
    }

    private boolean isNotActive(Cell cell) {
        return !isActive(cell.toString());
    }
}

File: src/main/kotlin/gameoflife/algorithm/GameOfLife.kt
package gameoflife.algorithm

class GameOfLife(private val livingCellsMap: Map<String, Cell>) {
    val livingCells get() = if (livingCellsMap.isNotEmpty()) livingCellsMap.values.toList() else emptyList()

    fun tick(): GameOfLife = GameOfLife(nextMap)

    private val nextMap: Map<String, Cell>
        get() =
            (nextGenerationCells + reproducibleCells)
                .distinct()
                .map { (it.toString() to it) }
                .toMap()

    private val nextGenerationCells get() = livingCells.filteredCells { it in 2..3 }
    private val reproducibleCells get() = activeNeighbours.filteredCells { it == 3 }
    private fun List<Cell>.filteredCells(filtered: (Int) -> Boolean) = filter { filtered(it.countActiveNeighbours) }

    private val String.isActive: Boolean get() = livingCellsMap[this] is Cell
    private val Cell.isInactive: Boolean get() = !toString().isActive
    private val Cell.countActiveNeighbours: Int get() = neighbours.filter { it.toString().isActive }.count()

    val activeNeighbours get() = livingCells.flatMap { it.neighbours }.distinct().filter { it.isInactive }

    fun isActive(x: Int, y: Int) = toName(x, y).isActive
}